Added lookAt matrix

bogdanteleaga opened the pull request:
So I was using your library in a project and I saw it lacks the lookAt matrix. After trying some other stuff I decided to add it myself. I've been using it for some days now with success, so I'd say it works properly. I also added the object wrapper since I mainly use them as objects. I had to modify my original implementation where I was using vec3's and mat4's with ^ and |, but I tested this out for a couple of minutes and it works.

Also, I've included a small driveby fix in the projection matrix where you weren't passing the dtype along.

bogdanteleaga commented:
Fixes 
https://github.com/adamlwgriffiths/Pyrr/issues/20

bogdanteleaga commented:
Renaming shouldn't be a problem. As for the tests, probably the best way to do it is to get numbers from a 'reliable' implementation, however I know none, since I found many different ones. 

What I can do it get some different outputs and plug them into the tests and just test that the operations are fine and it doesn't change over time.

Opinions?

adamlwgriffiths commented:
Just noticed that the `dtype` doesn't get passed used in `create_look_at`, the dtype is hard coded to 'f'.
```
(-np.dot(side, eye), -np.dot(up, eye), np.dot(forward, eye) , 1.0)), dtype='f')
```

Also `look_at` in OO API is forcing dtype to be None.
```
return cls(matrix44.create_lookAt(eye, target, up, dtype=None))
```

These need to pass the dtype parameter through.


I had exactly the same problem you mention regarding testing maths functions.
I've found the best way is to test the basic logical case for rotations, i.e. unit X rotate 1/2pi == -unit Z.

For more complex situations I've had to resort to wolfram alpha which can give some very good, specific, answers to questions.
Having a quick look, it doesn't seem to understand 'matrix look at' or similar search queries =/.

I think the best is still what I mentioned previously.
Create a look at matrix which looks, say, right (1/2pi rotation clockwise).
Apply to a unit x, unit y, and unit z vector.
Check that the unit vectors now face unit z, unit y, unit -x, respectively.
Then duplicate this for look up, etc.

From PyEuclid (https://code.google.com/p/pyeuclid/source/browse/trunk/euclid.py)

    def new_look_at(cls, eye, at, up):
      z = (eye - at).normalized()
      x = up.cross(z).normalized()
      y = z.cross(x)
      
      m = cls.new_rotate_triple_axis(x, y, z)
      m.d, m.h, m.l = eye.x, eye.y, eye.z
      return m

    def new_rotate_triple_axis(cls, x, y, z):
      m = cls()
      
      m.a, m.b, m.c = x.x, y.x, z.x
      m.e, m.f, m.g = x.y, y.y, z.y
      m.i, m.j, m.k = x.z, y.z, z.z
      
      return m

Using this code, we could create a matrix then compare against the result of pyrr's equivalent look at function.

I've noticed that there is a difference in how the X,Y,Z are computed.
Ours does `target - eye`, and PyEuclid does `eye - at`.
Perhaps the function accounts for this, as ours uses `-np.dot(...)` and theirs doesn't.
I think this shows that we need to do the unit x,y,z rotation to prove the function.
-Z is camera forward (right-handed coords), so applying a look at matrix to a vector will resolve this.

I should probably add that 'matrix from 3 vectors' function too =P.

Also, please add yourself to the authors list at the end of the README as part of the pull request =)

adamlwgriffiths commented:
cgkit also has a look at function, we can compare results against that.

http://cgkit.sourceforge.net/gallery.html
http://cgkit.sourceforge.net/doc2/lookat.html

bogdanteleaga commented:
Ok, I had another look at this. I fixed the problems you mentioned with dtype first of all as well as renaming the function.

As for testing, I tried installing cgkit, however it does not seem to work on my machine because of boost. The implementation in euclid is flat out wrong, there needs to be a translation followed by a rotation. The dot product takes care of that in one go, you can check the details here http://www.3dgep.com/understanding-the-view-matrix/. 

I also changed to a clearer implementation. I was doing target - eye, but when creating the matrix I was using -forward. Also, the dot product had it's sign changed. I switched to the implementation in the link - which is the same really - and it seems to work after some live testing. You can check it exactly in the latest commit.

Unfortunately, I didn't have time to add unit tests yet, hopefully this weekend.

adamlwgriffiths commented:
Sorry to be a pain, the Matrix object is still called `create_lookAt`.

dragostis commented on pyrr/matrix44.py:
```diff
@@ -341,6 +343,30 @@ def create_orthogonal_projection_matrix(
+    up = vector.normalise(np.cross(forward, side))
+
+    return np.array((
+            (side[0],            up[0],            forward[0],           0.),
```
Fugly syntax. PEP8 could be nice?

adamlwgriffiths commented:
Any progress on this?
If you can fixing the function name and adding a test I would be happy to merge it.

cprogrammer1994 commented:
will this be merged soon?

adamlwgriffiths commented:
For it to be merged he has to make the changes I stated:
1. Fix the function name so it's consistent with the rest of the library
2. Add tests

I'm not available for a few weeks. Perhaps I can do it myself after that.

cprogrammer1994 commented:
merged this branch to #47
